src/error.rs

```diff
@@ -139,3 +139,85 @@ where
 {
     DbErr::Json(s.to_string())
 }
+
+/// An error from unsuccessful SQL query
+#[derive(Error, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum SqlErr {
+    /// error for duplicate record in unique field or primary key field
+    #[error("Unique Constraint Violated")]
+    UniqueConstraintViolation,
+    /// error for Foreign key constraint
+    #[error("Foreign Key Constraint Violated")]
+    ForeignKeyConstraintViolation,
+}
+
+#[allow(dead_code)]
+impl DbErr {
+    /// Convert generic DbErr by sqlx to SqlErr, return none if the error is not any type of SqlErr
+    pub fn sql_err(&self) -> Option<SqlErr> {
+        #[cfg(any(
+            feature = "sqlx-mysql",
+            feature = "sqlx-postgres",
+            feature = "sqlx-sqlite"
+        ))]
+        {
+            use std::ops::Deref;
+            if let DbErr::Exec(RuntimeErr::SqlxError(sqlx::Error::Database(e)))
+            | DbErr::Query(RuntimeErr::SqlxError(sqlx::Error::Database(e))) = self
+            {
+                let error_code = e.code().unwrap_or_default();
+                let _error_code_expanded = error_code.deref();
+                #[cfg(feature = "sqlx-mysql")]
+                if e.try_downcast_ref::<sqlx::mysql::MySqlDatabaseError>()
+                    .is_some()
+                {
+                    let error_number = e
+                        .try_downcast_ref::<sqlx::mysql::MySqlDatabaseError>()?
+                        .number();
+                    match error_number {
```

Comment:
Hey @darkmmon, I wonder why for MySQL we use `error_number` to determine the error type instead of using `_error_code_expanded` just like other databases?

Reply:
@billy1624 
The reason is that the error code for MySQL is ambiguous. If I wanted to do it with .code() like for other databases, I would not be able to distinguish the violation between unique constraint and foreign key constraint (they both return 23000). 
I can only think of using the exact error number for distinguishing these errors, leading to the clumsy code. If there is an elegant way it would definitely be welcomed!